File: blind_75_leetcode/python/s0124_binary_tree_maximum_path_sum.py
from typing import Optional
import unittest


from .s0104_maximum_depth_of_binary_tree import TreeNode, create_tree_from_list


class Solution:
    def __init__(self):
        self.max_sum = float("-inf")

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def max_gain(node: Optional[TreeNode]) -> int:
            if node:
                left = max(max_gain(node.left), 0)
                right = max(max_gain(node.right), 0)
                self.max_sum = max(self.max_sum, left + right + node.val)

                return max(left, right) + node.val
            else:
                return 0

        max_gain(root)
        return int(self.max_sum)


class TestS124(unittest.TestCase):
    def test_works(self):
        self.assertEqual(Solution().maxPathSum(create_tree_from_list([1, 2, 3])), 6)
        self.assertEqual(
            Solution().maxPathSum(
                create_tree_from_list([-10, 9, 20, None, None, 15, 7])
            ),
            42,
        )

File: blind_75_leetcode/src/solution/s0124_binary_tree_maximum_path_sum.rs
#![allow(dead_code)]
pub struct Solution {}

use super::s0104_maximum_depth_of_binary_tree::TreeNode;

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn max_path_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut max_sum = i32::MIN;
        Solution::traversal(&root, &mut max_sum);
        max_sum
    }

    fn traversal(root: &Option<Rc<RefCell<TreeNode>>>, max_sum: &mut i32) -> i32 {
        if let Some(node) = root {
            let val = node.borrow().val;
            let left_gain = Solution::traversal(&node.borrow().left, max_sum).max(0);
            let right_gain = Solution::traversal(&node.borrow().right, max_sum).max(0);
            *max_sum = (val + left_gain + right_gain).max(*max_sum);

            left_gain.max(right_gain) + val
        } else {
            0
        }
    }
}

#[cfg(test)]
mod tests_124 {
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(
            Solution::max_path_sum(TreeNode::form_vec(vec![Some(1), Some(2), Some(3)])),
            6
        );
        assert_eq!(
            Solution::max_path_sum(TreeNode::form_vec(vec![
                Some(-10),
                Some(9),
                Some(20),
                None,
                None,
                Some(15),
                Some(7)
            ])),
            42
        );
    }
}
